### DISCUSSION TITLE
Entering magnitudes instead of fluxes into Phoebe?

### DISCUSSION THREAD

**Discussion by kaleetock** [phoebe-project/phoebe2#507] [Questions]
Hi! I am a new Phoebe user trying to enter magnitudes instead of fluxes into Phoebe using the following command: 

 b.add_dataset('lc', times=ts_2pv, magnitudes=mags_2pv, dataset='lc02')

This does not give an error, but then when I try to print them using

  print(b['magnitudes@lc02@dataset'])

it says "Parameter Set: 0 parameters" and when I try to plot them it tells me that there is nothing to plot.  

If I rename my "magnitudes" to "fluxes" in the data entry step, it will both print and plot my data (though in that case I get an error when I call run_compute).  Any wisdom would be greatly appreciated!

**Comment by kecnry** [comment]
Phoebe currently works strictly in fluxes, so you'll need to convert your magnitudes to fluxes before attaching the dataset.

Support for converting _within_ phoebe is on our list of feature requests, so might be available in a future release.  If you'd be interested in helping develop or test that, let us know!

**Reply by kaleetock** [reply]
Ok, I converted my calibrated magnitudes to calibrated fluxes (to within a scale factor.  I think.).  If I did that correctly, it still would only work for a Bessel-V filter.  If I confirm the conversion and find the corresponding conversions for other filters, I'll be happy to share.

However, in the meantime, I'm still struggling with Phoebe.  My bar is low: I just want to put in my data and get out something resembling the lightcurve I know is there.   I made a Colab at the link below.  What am I doing wrong?  Thanks so much for any wisdom!  

  https://colab.research.google.com/drive/18yYTiyjlhSojgELV_2PjIlPlQmdg8krn?usp=sharing

**Reply by kecnry** [reply]
A few suggestions:
* run_compute runs the _forward model_ only - it takes all the input parameters and creates a synthetic model.
* if you also pass [compute_times/compute_phases](http://phoebe-project.org/docs/latest/tutorials/compute_times_phases) when adding the dataset, then the synthetic model can be computed in evenly spaced time/phases rather than those matching the dataset
* the main difficulty you have in plotting the observations and the synthetic model is the [flux scaling](http://phoebe-project.org/docs/latest/tutorials/pblum).  My suggestion would be to use `pblum_mode='dataset-scaled'` for now, and later down the road consider if you want to actually fit for passband luminosities (which I would suggest if you eventually do MCMC).

You may also find some of the notebooks and recordings from the [2021 workshop](http://phoebe-project.org/workshops/2021june/materials) useful.  Just keep in mind some of these do use unreleased features (although we tried to state when that was the case) and are not kept up to date as the code changes going forward.  Since the workshop was so recent, it should still be quite current and you shouldn't really run into any changes yet.

**Reply by kaleetock** [reply]
Thank you so much for the links and workshop videos; these will keep me busy for a while!  :)

Could you let me know how to convert between W m^-2 and W m^-2 Hz^-1 ?  I gather one multiplies by a frequency, but *which* frequency?  

Thanks again for the references and resources!  :)

**Reply by mfitzasp** [reply]
We can convert magnitudes to Janskys and then Janskys to W/m2/Hz but we don't know how to go from W/m2/Hz to W/m2 for a given filter. It appears PHOEBE wants W/m2. Is it as simple as multiplying by the central frequency (1/Hz) of the filter?

**Reply by aprsa** [reply]
Let me try and elucidate this. PHOEBE works in fluxes to avoid problems with calibrations (see the [IAU resolution B2](https://arxiv.org/abs/1510.06262)). If you have magnitudes, you need two constants that are, fortunately, quite inconsequential unless you are trying to get distances figured out as well. The first is m<sub>0</sub>, the magnitude zero-point, and the second is f<sub>0</sub>, the flux that corresponds to the zero-point magnitude. In a nutshell, to convert from magnitudes to fluxes:

* if you have differential (relative) magnitudes, select m<sub>0</sub> for each passband to be close to the quarter-phase magnitude and then do f<sub>k</sub> = 10<sup>-2/5 (m<sub>k</sub>-m<sub>0</sub>)</sup> for all observed m<sub>k</sub>;
* if you have all-sky (absolute) magnitudes and do not care about absolute fluxes, select a single m<sub>0</sub> for all passbands that is close to a single passband's quarter-phase magnitude; then transform *all* magnitudes to fluxes using that same m<sub>0</sub> by the same equation as above;
* if you have all-sky (absolute) magnitudes and do care about absolute fluxes (i.e. to determine distance, reddening, etc), you need to calibrate m<sub>0</sub> and f<sub>0</sub> appropriately, which is outside the scope of PHOEBE. You need to concern yourself with the bolometric correction (cf. the resolution above) and the actual passband calibration, which is instrumental and had to have been used when calibrating magnitudes in the first place.

If you can provide me with your actual use case, I might be able to provide further guidance.

One last thought on the units: PHOEBE calculates *intensities* from each surface element, which are in W/m<sup>3</sup>, and then integrates them into fluxes, which are in W/m<sup>2</sup>, that are then compared to observations. Intensity is a local passband quantity that varies across the star; flux is a global passband quantity that varies with time.

**Comment by mfitzasp** [comment]
We do have absolute magnitudes. They are colour-term corrected such that the filters are very good representations of the standard filtersets and have very high confidence that they are accurate measurements of absolute magnitude on the standard scale. We have ugriz and BV filters, although at the moment we are playing around just with the B and V filter data to see if we can get PHOEBE to work. Due to the large number, and quality, of the filters and telescopes available for us to observe and we are currently using an Algol detached binary as a test object, it would be good to see if we could transform the absolute magnitudes into absolute fluxes to estimate the distance and reddening.

**Reply by aprsa** [reply]
Excellent! Do you have any standards observed? That will give you the needed m<sub>0</sub>-f<sub>0</sub> transformation. You will obviously need the standards observed in each passband that you want to calibrate, preferably observed in succession with your binary.

If you don't have any standards at hand, then perhaps dig into the reduction pipeline. The data from the CCD are photon counts, which is flux; those numbers were then converted to absolute magnitudes using *some* standards, or perhaps a known m<sub>0</sub>-f<sub>0</sub> relationship. If you can get those, your job is done -- just "undo" the magnitude conversion, leave your data in W/m^2 and phoebe will adjust its passband luminosity (or distance, depending on the mode of operation) to match your absolute units.

Please let us know if we can help with any of this.

**Reply by mfitzasp** [reply]
Hi! We have standards within frame, usually from APASS or SDSS or Skymapper (depending on which object we are looking at). While they are secondary standards (rather than primary landolt/clem/etc standards), they are numerous and the within frame variation is tiny compared to between frames and we get very good correlation between catalogue and measured values (around 0.01 mags). We have found, easily, the conversion from magnitudes to Janskys (or W/m^2/Hz) to get the calibrated flux. What we are not understanding is how you go from W/m^2/Hz to W/m^2 .... is the Hz part a formality? It is just saying "at this central wavelength of the filter, it is this W/m^2" or do we have to multiply by a value with the units of Hz to cancel it out (e.g. the central wavelength). So we are very close to having calibrated fluxes, our problem is with the units required for PHOEBE. Jansky =/= W/m^2

**Reply by aprsa** [reply]
I'm not sure how you go from magnitudes (a measure of flux) to janskys (a measure of flux density)? My guess is that you're treating your passbands as monochromatic at a central frequency. That will likely cause quite a bit of error compared to what phoebe does (integrates over the actual passband response). My advice would be convert magnitudes to fluxes as that requires no assumptions of any kind (other than the m0/f0 calibration, of course).

**Reply by mfitzasp** [reply]
https://ui.adsabs.harvard.edu/abs/1998A%26A...333..231B/abstract

![image](https://user-images.githubusercontent.com/16806710/128938222-886ba157-c1b6-426c-a180-ad4b3dc90782.png)

Basically using something like that. This is just one example, but this one is in ergs /cm^2/sec/Hz. Most give it in Janskys, others give it in W/m^2/Hz .... none give it in flux, all give it in flux density. So all of the conversions given by calibration sites are flux density. How do we get it in flux?

other examples:

SDSS: http://classic.sdss.org/dr3/algorithms/fluxcal.html#sdss2flux

http://www.astronomy.ohio-state.edu/~martini/usefuldata.html

Hubble: https://www.stsci.edu/hst/instrumentation/wfc3/data-analysis/photometric-calibration

Pretty sure Skymapper is the same, but I can't find the reference off-hand.

**Reply by aprsa** [reply]
As best as I can tell, these all give *equivalent* flux densities, i.e. what the flux of a *uniform* (flat) passband response would be for some predefined SED. For example, from the last reference you provided, "The STmag and ABmag systems define an equivalent flux density for a source, corresponding to the flux density of a source of predefined spectral shape that would produce the observed count rate, and convert this equivalent flux to a magnitude." To first order that's of course fine, but that neglects all passband effects (i.e., dependence of effective passband wavelength on the SED itself). If that's good enough, then taking spectral density and integrating it over the wavelength range of the passband will give you flux, i.e., Fpb = dF/dlambda (lambda_hi - lambda_lo), where dF/dlambda is the flux density adopted from the literature, lambda_lo is the shortest wavelength at which dF/dlambda is non-zero, and lambda_hi is the longest wavelength at which dF/dlambda is non-zero. Same for frequencies, of course, I just tend to think in terms of wavelengths, and in terms of pre-integrated passbands.

**Reply by aprsa** [reply]
And just to clarify: by uniform passband I don't mean to imply that the passband itself is uniform, but that the integral over its shape is taken to be independent of the shape itself: \int_\lambda S(\lambda) P(\lambda) d\lambda / \int_\lambda P(\lambda) d\lambda for some predefined S(\lambda).

**Reply by mfitzasp** [reply]
Okiedokie.... this is well worth your time :) I'm hoping to synthesize this into a "how to get stuff into peranso for observers" type doc... I also work better in wavelengths....

So if we have a df/dlambda = 3631 (I am just picking a random number here - it isn't physical, just for example.)

then Fpb = 3631 * (lambda_hi - lambda_lo)

So given a passband curve for the standard V that I got offline (lets assume it is correct for arguments sake): https://www.dropbox.com/s/lehht45itdt6r69/Bessel_V-1.xlsx?dl=0 (adapted from https://old.aip.de/en/research/facilities/stella/instruments/data/johnson-ubvri-filter-curves)

So looking at this passband, then the lowest lambda is about 464 nm and the highest nm is about .... well.. there is a long tail, but lets say 715 nm.

So then, assuming all the units are correct, then flux would be:

Fpb = 3631 * (465 nm - 715 nm)

but this seems to be highly dependant on where you set the lowest lambda and the highest lambda... which itself is not absolutely clear.... And wouldn't that be significantly different than the real passband? 

![image](https://user-images.githubusercontent.com/16806710/128953592-2658f33c-87ab-41ae-ac1a-6fa60da33d06.png)

so shouldn't we instead take the actual passband and then sum up all the little bits of the passband? Taking into count units as per usual, we should multiply the small steps of the passband throughput data by the flux density and then sum up these individual elements to get the area underneath the curve? And that gives us the flux? 

![image](https://user-images.githubusercontent.com/16806710/128953903-60ab3b1f-f36d-43c1-91c6-6030e88b79f6.png)

It would seem if that was the case, then for each filter IF you assume that it matches closely to a standard filter and IF you assume the star you are looking at isn't so strange it introduces differences, then this would be a standard multiplicative value for each standard filter to multiply the flux density to get the flux.... unless I am missing something?

**Reply by aprsa** [reply]
And the latter is *precisely* how phoebe does it under the hood, but you need an actual SED for that. Here are the [details](http://aprsa.villanova.edu/files/radiation.pdf), Table 4.2 in particular. The first bit to understand at this point is exactly how skymapper (or whatever was used to get the magnitudes) calibrates its magnitudes, which I'm still fuzzy about.

**Reply by mfitzasp** [reply]
Well, the main focus we have here is how to take reliably calibrated magnitudes on a standard scale (whether that be ugriz or BV from APASS, Skymapper, SDSS, Landolt or wherever) to something that is importable to PHOEBE as fluxes (as opposed to flux densities). If we have reliable calibrated magnitudes (derived from our observations) then we should be able to supply reliable calibrated fluxes in whatever PHOEBE would like to receive.

Is your question about how the original Skymapper catalogue got it's initial magnitudes? That is a big question. Skymapper griz was tied to Panstarrs-1 griz. 

Or how we derived our calibrated magnitudes from the Skymapper catalogue? The short version of this is we got our calibrated photometry from comparison to an ensemble set of comparison stars in the frame. We have a colour-term correction element, but the colour-term is very small usually. The match to the original catalogue (Skymapper or APASS or SDSS or whatever) is usually acceptable small - 0.01 standard deviation when comparing measured calibrated magnitudes against the original catalogue. 

If we are talking B or V, then at this stage we are speaking of APASS (https://www.aavso.org/apass). That is tied to the original Landolt & Clem primary standards on the Vega system is my understanding. 

It would seem that a B and a V calibrated lightcurve of this nature would be the the most common two bits of data an observer might want to bring into PHOEBE. I had done this previously many moons ago in the original PHOEBE.

**Reply by mfitzasp** [reply]
More directly... here is an actual dataset. BJD and Vmag and Verror observed. Next column the 0.02 subtraction to get it onto the AB scale. Then from there to Fv in Janskys (using m(AB) = -2.5 log10(fv) + 8.90). Then a unit conversion from Janskys to w/m^2/Hz (* 10^-26). Knowing this is observed using a standard V filter, how do we jump from there to W/m^2 to provide this data to PHOEBE?

[Vdataset.xlsx](https://github.com/phoebe-project/phoebe2/files/6965462/Vdataset.xlsx)

**Reply by mfitzasp** [reply]
Hi! 

I was wondering if you had thought more about this? It seems like a fundamental thing that PHOEBE would need to do - take observed calibrated magnitudes in as observed calibrated fluxes. It is a nice learning experience about how to do this manually - and it is perhaps a good instructive thing to do as well - but what is the missing step or concept that we are missing here?

Michael

**Reply by kecnry** [reply]
We are planning to at least implement helper function which would take magnitudes, m0, and f0, and return fluxes which can then be input into phoebe (and the opposite should you want to export synthetic magnitudes from phoebe fluxes).  There is a chance this might be implemented as an optional constraint, but that hasn't been decided quite yet (it does add a bit of extra complications, but comes with the advantage of allowing plotting magnitudes directly in phoebe).  Any more advanced conversion to/from flux densities would still need to be done manually, although I would suggest avoiding it altogether unless absolutely necessary.

**Reply by aprsa** [reply]
@mfitzasp, we're wrapping up a major TESS EB effort that's taking all of my time; I'll get back to this once I can actually put the data under my fingertips. Stay tuned, I'll try to get to this as soon as practical.

**Reply by mfitzasp** [reply]
No problem, thank you very much! Looking forward to it! It would be nice to get my head around the conversions (even if the helper function makes it realtively straightforward anyway).

Thanks for your help!

**Reply by mfitzasp** [reply]
Hope you are doing well, I was wondering if you had thought about this more in the intervening handful of months :)

**Reply by aprsa** [reply]
Thanks for the ping on this -- I'll refresh my memory next week and see if I can come up with something coherent soon.

**Reply by mfitzasp** [reply]
No problem! Thanks for taking a look :)

**Reply by mfitzasp** [reply]
Hi! I was wondering if you'd had a chance to look at this? I appreciate there are a hundred other higher priorities though :)

**Comment by aprsa** [comment]
I've been on this for the last week and I have a jupyter notebook essentially ready that demonstrates all the points I raised earlier. It is quite intricate and I will try to briefly summarize it here, and -- once the notebook is polished -- I'll send out the link to it; I just didn't want to leave this hanging even longer than it already has been. We're preparing a workshop that starts in less than a month so time is a commodity in short supply these days. Also note that there's a [bug](https://github.com/github/markup/issues/1546) with latex markup that doesn't render some of the equations below; hopefully they resolve it soon and markup starts to work automagically.

- The measured quantity on a photometric CCD telescope system is passband *flux* (as opposed to flux density). In particular, it is the count of detected photons per second in a given amount of time. This photon count will depend on all instrumental aspects of the observational setup: filter transmission, optics, CCD QE, ... These are instrumental passband fluxes and they differ from site to site.
- Next, we need to reconcile site differences. For that we observe standard stars and we measure their own instrumental passband fluxes. For each standard we have a published value of the corresponding flux density that we will have to match. In order to do that, we need to calculate *instrumental flux density* of the calibration star. *How* we do that depends on the processing pipeline. I would argue that the best way would be to do this:

$$ f_\lambda^\mathrm{instrumental} = \frac{\int_\lambda \lambda I(\lambda) P(\lambda) d\lambda}{\int_\lambda \lambda P(\lambda) d\lambda} = \frac{f_\mathrm{pb}^\mathrm{instrumental}}{\int_\lambda \lambda P(\lambda) d\lambda}, $$

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;where $I(\lambda)$ is the (unknown) spectral energy distribution of the observed calibration source and $P(\lambda)$ is the passband response.

- This instrumental flux density is *passband-independent* -- it is a constant value across the entire wavelength span. If integrated over the passband-weighted response, it would yield the same integral as the source. We can now calibrate it: we know what $f_\lambda$ of the calibration source should be (for example, from the [STmag calibration](https://docs.astropy.org/en/stable/api/astropy.units.function.logarithmic.STmag.html)), so we have a simple scaling factor that maps $f_\lambda^\mathrm{instrumental} \mapsto f_\lambda$ that we can apply to the actual source of interest.
- We now have calibrated flux density $f_\lambda$. But this is flux *density*, not flux, and we need to convert it back to passband flux so that PHOEBE can use it. Thus:

$$ f = \frac{1}{hc} \int_\lambda \lambda f_\lambda P(\lambda) d\lambda = f_\lambda \left( \frac{1}{hc} \int_\lambda \lambda P(\lambda) d\lambda \right) \equiv f_\lambda A_\mathrm{photon}. $$

- So how do magnitudes come into all of this? The exact same way. We either refer to passband magnitudes or magnitude densities (more frequently known as monochromatic magnitudes). They are computed from:

$$ m - m_0 = -\frac 52 \log \frac{f}{f_0},  $$

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;where $m_0$ and $f_0$ are calibration constants (for example, $f_0 = 1 \, \, \mathrm{erg}/\mathrm{cm}^2/\mathrm{s}/\overset{\circ}{A}$ corresponds to $m=-21.1$ for the [STmag](https://docs.astropy.org/en/stable/api/astropy.units.function.logarithmic.STmag.html) system).

- In order to properly convert magnitudes to fluxes, we thus need to know how the conversion from observed fluxes to flux densities was done under the hood. Here I covered what I think is the most intuitive way. Other ways include bolometric weighing (i.e. energy- rather than photon-weighted), effective wavelength correction, etc. Then we simply need to *undo* that and we will get proper fluxes that PHOEBE can work with.

Thank you for bearing with me with the delay in responding. This semester has been just crazy...

**Reply by kaleetock** [reply]
Thank you so much for this comprehensive response.  The latex mostly works on Github where I am reading this, so all is well.  My semester has been crazy too, and I'm looking forward to digesting the explanations above once it is properly finished.  (One of these days I'll make it to the summer workshop--this summer doesn't work, unfortunately.)

Thanks again so much!

**Comment by aprsa** [comment]
In the hope it helps, [here is a notebook](https://github.com/phoebe-project/phoebe2-workshop/blob/master/2022june/Tutorial_flux_calibration.ipynb) that should elucidate some of the finer points of flux calibration. I'd love to receive any and all feedback on it.

**Reply by kaleetock** [reply]
Thank you so much for the notebook!  I appreciate the graphs and can tell that it will be helpful once I dig into it.  :)
